### ISSUE TITLE
[REVIEW]: Predihood: an open-source tool for describing and predicting neighbourhoods' environment

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#2805]
**Submitting author:** @fduchatea (<a href="http://orcid.org/0000-0001-6803-917X">Fabien Duchateau</a>)
**Repository:** <a href="https://gitlab.com/fduchate/predihood" target ="_blank">https://gitlab.com/fduchate/predihood</a>
**Version:** v1.1
**Editor:** @galessiorob
**Reviewer:** @jdalzatec, @omshinde, @nuest, @martinfleis
**Archive:** <a href="https://doi.org/10.5281/zenodo.4737729" target="_blank">10.5281/zenodo.4737729</a>

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/14eee5164e9b664fcfe62550b6924242/status.svg)](https://joss.theoj.org/papers/14eee5164e9b664fcfe62550b6924242)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/14eee5164e9b664fcfe62550b6924242"><img src="https://joss.theoj.org/papers/14eee5164e9b664fcfe62550b6924242/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/14eee5164e9b664fcfe62550b6924242/status.svg)](https://joss.theoj.org/papers/14eee5164e9b664fcfe62550b6924242)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://gitlab.liris.cnrs.fr/fduchate/predihood" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@jdalzatec & @omshinde & @nuest & @martinfleis, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @galessiorob know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://gitlab.liris.cnrs.fr/fduchate/predihood">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@fduchatea) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://gitlab.liris.cnrs.fr/fduchate/predihood">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@fduchatea) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://gitlab.liris.cnrs.fr/fduchate/predihood">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@fduchatea) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://gitlab.liris.cnrs.fr/fduchate/predihood">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@fduchatea) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @jdalzatec, @omshinde, @nuest, @martinfleis it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1109/socialcom.2013.17 is OK
- 10.5220/0009885702940301 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by galessiorob** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02805/joss.02805/10.21105.joss.02805.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02805/joss.02805/10.21105.joss.02805.pdf) :page_facing_up: :point_left:

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02805/joss.02805/10.21105.joss.02805.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02805/joss.02805/10.21105.joss.02805.pdf) :page_facing_up: :point_left:

**Comment by galessiorob** [comment]
👋 @omshinde, @nuest and @jdalzatec

Thank you all for volunteering as reviewers for this paper! At the top, you'll find individual checklists to work trough, please let me know if something is not clear or if you need any help.

**Comment by nuest** [comment]
Thanks! Just as a quick heads up, I'll likely do the review beginning of next week.

**Comment by fduchatea** [comment]
Thank you reviewing our submission, and sorry for the delay.

About mentioned missing items:
- _Permit individuals to create issues/file tickets against your repository_: it seems that the gitlab instance of our lab requires authentification (of lab members) for creating issues, and this policy will not change. Should we switch to another platform such as gitlab.com? 
- _Expand the description of the software_: Should we expand the description here, in README or in the paper?
- _Expand and focus the software's research scope_: Neighbourhoods are a very common concept in studies from diverse domains such as health, social sciences, or biology. For instance, 
Japanese researchers investigated the relationships between social factors and health by taking into account not only behavioural risks, but also housing and neighbourhood environments [1]. In a British study, authors describe how living areas have an impact on physical activities, from which they determine a walkability index at the neighbourhood level for improving future urban planning [2]. Smarts cities also consider neighbourhoods as an ideal unit division for measuring urban quality [3]. Lastly, a survey describes the luxury effect, i.e., the impact of wealthy neighbourhoods on the surrounding biodiversity [4]. However there is no clear definition of the neighbourhood environment. 
Our tool fills this gap by defining neighbourhoods and their environment, characteristics of these neighbourhoods and an interface for using popular machine-learning algorithms. These elements can be extended/enriched.
Our tool has been currently used to measure the impact of the neighbourhood's environment when people moves in another city [5]. But it could be extended to other application domains: measuring the pollution degree in neighbourhoods, determining whether a neighbourhood is suitable as stopover for migratory birds, etc.
We can reformulate and add this research scope in the paper if needed (we are already above the 1,000 words limit though).

[1] Takada,M.,Kondo,N.,Hashimoto,H.:Japanesestudyonstratification,health,income,and neighborhood: study protocol and profiles of participants. Journal of epidemiology 24(4), 334–344 (2014)
[2] Frank, L.D., Sallis, J.F., Saelens, B.E., Leary, L., Cain, K., Conway, T.L., Hess, P.M.: The development of a walkability index: application to the neighborhood quality of life study. British journal of sports medicine 44(13), 924–933 (2010)
[3] Garau, C., Pavan, V.M.: Evaluating urban quality: Indicators and assessment tools for smart
sustainable cities. Sustainability 10(3), 575 (2018)
[4] Leong, M., Dunn, R.R., Trautwein, M.D.: Biodiversity and socioeconomics in the city: a review of the luxury effect. Biology Letters 14(5), 20180082 (2018)
[5] Barret,N.,Duchateau,F.,Favetta,F.,Bonneval,L.:Predictingtheenvironmentofaneighbor- hood: a use case for france. In: International Conference on Data Management Technologies and Applications (DATA). pp. 294–301. SciTePress (2020)

**Comment by whedon** [comment]
:wave: @jdalzatec, please update us on how your review is going.

**Comment by whedon** [comment]
:wave: @omshinde, please update us on how your review is going.

**Comment by whedon** [comment]
:wave: @nuest, please update us on how your review is going.

**Comment by whedon** [comment]
:wave: @martinfleis, please update us on how your review is going.

**Comment by martinfleis** [comment]
It will likely take some time before I'll manage to do my review. Hard to estimate now, I haven't properly looked into the complexity of the package yet.

**Comment by omshinde** [comment]
>  @omshinde, please update us on how your review is going.

It's coming along nicely. I have started reviewing it locally based on the checklist, will keep posted via updating the checklist. Thanks!

**Comment by jdalzatec** [comment]
@whedon   I am playing a bit with the package while going through the checklist. I'll take some more time while reviewing it locally. Thanks!

**Comment by whedon** [comment]
I'm sorry human, I don't understand that. You can see what commands I support by typing:

```
@whedon commands
```

**Comment by galessiorob** [comment]
@jdalzatec @martinfleis @omshinde thank you all for the updates, please let me know if you need my help in any way.

@fduchatea thanks for the update on your part, please see my answers below:

> Permit individuals to create issues/file tickets against your repository: it seems that the gitlab instance of our lab requires authentification (of lab members) for creating issues, and this policy will not change. Should we switch to another platform such as gitlab.com?

To avoid having you migrate everything to another repo I propose two options:
 1. That you open a "review repo" in GitLab that allows us to reference the repo we are reviewing so that the reviewers can link to the specific files they on issues.
2. That you open a "review repo" in GitHub to do the same as above, however, I think that having the review repo and the actual repo on the same platform would be easier.

>Expand the description of the software: Should we expand the description here, in README or in the paper?

In the paper.

>Expand and focus the software's research scope: Neighbourhoods are a very common concept in studies from diverse domains such as health, social sciences, or biology. For instance, Japanese researchers investigated the relationships between social factors and health by taking into account not only behavioural risks, but also housing and neighbourhood environments [1]. In a British study, authors describe how living areas have an impact on physical activities, from which they determine a walkability index at the neighbourhood level for improving future urban planning [2]. Smarts cities also consider neighbourhoods as an ideal unit division for measuring urban quality [3]. Lastly, a survey describes the luxury effect, i.e., the impact of wealthy neighbourhoods on the surrounding biodiversity [4]. However there is no clear definition of the neighbourhood environment.
Our tool fills this gap by defining neighbourhoods and their environment, characteristics of these neighbourhoods and an interface for using popular machine-learning algorithms. These elements can be extended/enriched.
Our tool has been currently used to measure the impact of the neighbourhood's environment when people moves in another city [5]. But it could be extended to other application domains: measuring the pollution degree in neighbourhoods, determining whether a neighbourhood is suitable as stopover for migratory birds, etc.
We can reformulate and add this research scope in the paper if needed (we are already above the 1,000 words limit though).

I think we can just refactor the introduction and the statement of need to reflect the research need and scope by including some of the applications that you listed above. Could you take a first pass at it and I can help refine after that?

**Comment by fduchatea** [comment]
@galessiorob 

Thanks for the hints. We have created a repository with a public tracker:
https://gitlab.com/fduchate/review-repo-predihood

The first sections of the paper have been reformulated to broaden the description.

**Comment by fduchatea** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02805/joss.02805/10.21105.joss.02805.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02805/joss.02805/10.21105.joss.02805.pdf) :page_facing_up: :point_left:

**Comment by martinfleis** [comment]
I know it is not my role to judge this but I believe that JOSS's requirement that the software must "_Permit individuals to create issues/file tickets against your repository_" is not related to review process but to the actual software repository. Therefore a review only solution is not enough. The reason why I think so is that a software in a private, although readable repository is not entirely open in a community sense. Using [Matt Rocklin's seven stages of open software](https://coiled.io/blog/stages-of-openness/) `predihood` is currently on the stage 2, while my feeling about JOSS is that it requires stage 3. 

But again, it is up to editorial team of JOSS to judge this.

**Comment by nuest** [comment]
@martinfleis That's my understanding, too. https://gitlab.liris.cnrs.fr/fduchate/predihood/issues requires me to register/sign-in with a LIRIS account, which I, or any other external users, do not have.

@fduchatea Would you be willing to switch your "main" repository to a more open platform?

@galessiorob Please advise.

**Comment by nuest** [comment]
**Some first review comments and questions** I stopped the review at this point as some things seem to be missing (test files, clarification of authorship, public issue tracker). I invite the authors to respond and kindly ask to revisit the JOSS guidelines and the review checklist.

------

Note to self:

```Python
source .venv/joss-predihood/bin/activate
```

------

### General checks

- Contribution and authorship: there are three authors on the paper, but only two contributors on the repo - please clarify
- I'm hesitant to check of the "Substantial scholarly effort" as defined [here](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort). Has the software been used in other use cases than your own yet? I understand it makes addressing research challenges potentially easier, but I'd like to understand better how other researchers can load their own data into predimap, and how they can get their data _out_ of the system again.
  - The example you provide (find a new place where to live) is interesting, but not a scientific application IMO. What would be a showcase that answers a research question?

### Functionality

I was not able to complete all steps in the README, please @fduchatea clarify - thanks!

- I had to also run `pip install wheel` in my virtual environment to install predihood
- Waht about making the tool easier to try out by providing a Docker image? Or you could provide instructions for using MongoDB inside a container (which I prefer)
  - Note to self: I ran a MongoDB 4 container, `docker exec`ed into it, installed wget, download database dump, ran mongorestore
- I was not able to locate `dump-iris.bin` locally after installation. I did find the file `dump-dbinsee.bin` after manually downloading the `mongiris-master.zip` from https://gitlab.liris.cnrs.fr/fduchate/mongiris - _Is this the file?_
- On the map, I can calculate classifications for some areas, but not for all (sometimes nothing happens) - is that expected? If yes, it should be communicated more clearly to the user.
- Starting from the map: How can I export the calculated values? How can I compare the results of the different algorithms? Do I have to manually copy them to an external tool/spreadsheet? (Sorry if I'm missing something.

### Documentation

- Please expand in the README how to _run_ the tests (not just mention the file name); also, the file `tests.py` seems to be missing anyway.
- I suggest to mention the license in the README - not everyone will know where to look for it
- In the paper, you mention that data is stored as GeoJSON, and you mention other use cases - can you document these other use cases and how to load data besides the IRIS data (which is a MongoDB data dump and not a GeoJSON file) ?
- Community guidelines are missing

**Comment by galessiorob** [comment]
@martinfleis and @nuest, thank you both for your thoughtful comments and assessments on the current state of accessibility of this software. Our [submission requirements](https://joss.readthedocs.io/en/latest/submitting.html#submission-requirements) state that the software must: 

- Be stored in a repository that can be cloned without registration.
- Be stored in a repository that is browsable online without registration.
- Have an issue tracker that is readable without registration.
- Permit individuals to create issues/file tickets against your repository.

@fduchatea can you change the settings on the repo so they comply with the above, and we'd also need this repo to be the source one - not the one that's restricted to having a signing for LIRIS.

CC @arfon @danielskatz

**Comment by galessiorob** [comment]
@fduchatea checking in on the aforementioned requirements, want to make sure you'll be able to provide them so we can carry on with the review. Let me know if you have any questions!

**Comment by fduchatea** [comment]
@galessiorob we understand that it is an issue not to let individuals create tickets. We have migrated to a more open platform and we are discussing about mentioned requirements.
https://gitlab.com/fduchate/predihood

**Comment by galessiorob** [comment]
Thanks so much @fduchatea

**Comment by fduchatea** [comment]
@galessiorob @nuest

As previously mentioned, the new repository is now on https://gitlab.com/fduchate/predihood for better accessibility.

Following are answers to most of the mentioned points.

- Contribution and authorship: there are three authors on the paper, but only two contributors on the repo - please clarify
```
This project was implemented during a 6 months training period. Nelly B. was supervised by F. Favetta and F. Duchateau.
Although F. Favetta has not directly performed commits, he actively participated in the supervision of Nelly, definition and ideas about algorithms and implementation details, testing/installing of Predihood, writing/reviewing the paper.
```

- I'm hesitant to check of the "Substantial scholarly effort" as defined here. Has the software been used in other use cases than your own yet? I understand it makes addressing research challenges potentially easier, but I'd like to understand better how other researchers can load their own data into predimap, and how they can get their data out of the system again. The example you provide (find a new place where to live) is interesting, but not a scientific application IMO. What would be a showcase that answers a research question?
```
Predihood is the result of a 6 months work (February til August 2020) for 3 persons. Besides, it reuses mongiris (a light API for interacting with the neighbourhood database) which still required a few months of work.
As discussed with galessiorob, the tool has currently been only used in our context (tool developed and paper published in Summer 2020, which also limits its visibility). 
We have added examples of potential applications in the paper: measuring the pollution degree in neighbourhoods, determining whether a neighbourhood is suitable as stopover for migratory birds, etc.
We are developing another (small) use case to explain how to load other data.
We have added an export functionality for all neighbourhoods selected on the map. In our initial context, we had to copy the results (the search was limited to several neighbourhoods) but we agree that prediction from the map should have an export functionality.
```

- What about making the tool easier to try out by providing a Docker image? Or you could provide instructions for using MongoDB inside a container (which I prefer)
```
We do not have any expertise on Docker, so we will probably not be able to take into account this point in reasonable time.
Maybe some contributors may be willing to add this feature later?
```

- I was not able to locate dump-iris.bin locally after installation. I did find the file dump-dbinsee.bin after manually downloading the mongiris-master.zip from https://gitlab.liris.cnrs.fr/fduchate/mongiris - Is this the file?
```
Yes, this is the database file. We have updated the README.
```

- On the map, I can calculate classifications for some areas, but not for all (sometimes nothing happens) - is that expected? If yes, it should be communicated more clearly to the user.
```
We were not able to reproduce this bug. We have tested hundred of neighbourhoods, but as there are 50,000 in the database, we.
Do you have some neigbourhoods name or code (which do not produce any prediction) so that we can investigate?
```

- Starting from the map: How can I export the calculated values? How can I compare the results of the different algorithms? Do I have to manually copy them to an external tool/spreadsheet? (Sorry if I'm missing something.
```
You are right, we used to manually copy results when predicting on the map. We have adding an export functionality on the cartographic interface (XSL file export).
In the training part, there is an export function of the results.
```

- On the classifier training part. When I select both "Remove outliers" and "Remove rural", I get errors in the console but only a generic error message in the UI - more helpful error messages for the users would be important. Below, the problem is that the test size is negative, but the message is generic
```
Error when selecting both outliers and rural is fixed (issue due to Python 3.8).
Specific error messages have been added in the UI.
```

- Please expand in the README how to run the tests (not just mention the file name); also, the file tests.py seems to be missing anyway.
```
We updated the path to the test file (predihood/tests.py) and we indicated how to run them.
```

- I suggest to mention the license in the README - not everyone will know where to look for it
```
Added.
```

- I had to also run pip install wheel in my virtual environment to install predihood
```
We have tested in a new (clean) virtual environment, and there was no need to install wheel.
According to documentation and Stackoverflow, it seems that wheel is packaged automatically with latest versions of pip (above 19.2). 
```

We are working on the last points (loading data and community guidelines - although we are not sure what to write for this latter).

**Comment by nuest** [comment]
Thanks for the updates so far!

To me making clear how the software is useful and usable (data import/export) to others is key, because if it is limited to your own research, I personally think the corresponding scientific article is a suitable way to reference it. Just wanted to put that here as this concern might not have been clear in my previous comments.

Re. community guidelines: take a look at recent published JOSS papers and I'm sure you'll find some good examples.

_I'll wait for you to complete your changes and will revisit my review then._

**Comment by martinfleis** [comment]
Just a ping here - I have opened an issue https://gitlab.com/fduchate/predihood/-/issues/1 since the installation following the instructions failed in my case. Therefore I am waiting for this to be resolved.

**Comment by fduchatea** [comment]
We are refactoring the code to facilitate import of new data.
We will check the issue later.

**Comment by galessiorob** [comment]
Hi @fduchatea! I hope you had a nice holiday break 🎄 

Checking in on several things so the reviewers can keep doing their work; some of the major issues to work on from your side:
- [ ] [Installation instructions issue](https://gitlab.com/fduchate/predihood/-/issues/1) (kindly opened by @martinfleis)
- [ ] [Clarifying data import and export usability](https://github.com/openjournals/joss-reviews/issues/2805#issuecomment-744361435)
- [ ] Export functionality on the cartographic interface (XSL file export)
- [ ] Data and community guidelines. Some resources for this:
     - [Setting up your project for success with community guidelines](https://opensource.guide/building-community/)
     -  [Open Source Code of Conduct](https://opensource.guide/code-of-conduct/)
     - [Intro blog for the GitHub Community guidelines](https://github.blog/2016-10-19-introducing-github-community-guidelines/) (dated but the links lead to the current resources)
     - [GitHub Community guidelines, as an example](https://docs.github.com/en/free-pro-team@latest/github/site-policy/github-community-guidelines)
     - [Google OS Community Guidelines](https://opensource.google/conduct/)
     - [Code of conduct examples](https://indieweb.org/code-of-conduct-examples)
     - [One more example from a smaller project that I think is well written](https://github.com/eiffel-community/.github/blob/master/CODE_OF_CONDUCT.md)

Let me know if I can help clarify any of this!

**Comment by fduchatea** [comment]
Hi @galessiorob 
Thank you, and we wish an Happy New Year to all of you.

We have added a new (fake) dataset (prediction works on it), a CSV export functionality and the Community guidelines (based on the links you provided, thanks). The README has been updated to describe how to use/import another dataset.

The updated code is still on a _dev_ branch, we still perform some tests).
We should merge the branch at the beginning of next week and will tell you.

**Comment by galessiorob** [comment]
Thanks, @fduchatea !

**Comment by fduchatea** [comment]
We have merged into the master branch.
We are still correcting a few bugs, but it is possible to load and predict for another dataset.
The README has been updated to reflect these changes.

**Comment by galessiorob** [comment]
Thank you @fduchatea 

@jdalzatec, @omshinde, @nuest, @martinfleis please resume your reviews at your earliest convenience 🙏 
Let me know if there's anything I can help with.

**Comment by fduchatea** [comment]
New bugs have been corrected this week:
- fix CSS footer issue
- fix error when using the example predictive algorithm MyOwnClassifier (implemented basic code to solve this issue)
- fix error "object internal error" when predicting in a popup using MyOwnClassifier
- fix the error 404 when displaying more-details page (for bird-migration dataset)
- fix issue with the number of splits in the testing interface
- fix some minor GUI issues

**Comment by nuest** [comment]
@fduchatea Are you currently actively developing and fixing bugs? I don't want to play catch with the developments for the review. (AFAIK, the JOSS submission should happen for rather stable pieces of software.)

**Comment by fduchatea** [comment]
@nuest No we are not in active development. We refactored the code 2 weeks ago to facilitate the use of new datasets (code merged around Jan 20th) and we are not developing anymore. 
Last week we noticed a few bugs so we corrected them, but the application is usable.

**Comment by galessiorob** [comment]
@jdalzatec, @omshinde, @nuest, @martinfleis I recognize that this paper has elongated past the ideal time, if you can, please resume your reviews. Let me know if you have a conflict.

Thanks!

**Comment by jdalzatec** [comment]
@galessiorob Hi Gabby, I will check it ASAP. Thanks!

**Comment by omshinde** [comment]
Hi @galessiorob ! Thanks for the reminder. I will submit my reviews as early as possible.

**Comment by martinfleis** [comment]
The major blocker on my side is the installation (xref https://gitlab.com/fduchate/predihood/-/issues/1). I was not able to install and run `predihood` so far despite several attempts. The application of this kind would hugely benefit from a docker container which would directly start the app on `docker run`.

**Comment by omshinde** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
PDF failed to compile for issue #2805 with the following error: 

 Can't find any papers to compile :-(

**Comment by arfon** [comment]
@whedon generate pdf

**Comment by arfon** [comment]
@galessiorob - I updated the repository address above to point to https://gitlab.com/fduchate/predihood (I *think* that's correct)

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02805/joss.02805/10.21105.joss.02805.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02805/joss.02805/10.21105.joss.02805.pdf) :page_facing_up: :point_left:

**Comment by galessiorob** [comment]
@arfon thank you!

**Comment by omshinde** [comment]
Hi @fduchatea and authors!
cc: @galessiorob 

System settings while reviewing:

    Installation - Local Machine
    OS - Ubuntu 20.04 (WSL)
    Python version - 3.8.5

Thank you for submitting a wonderful application, I foresee Predihood's utility as a great application for predicting insights about a (the) neighborhood (s). Please find my thoughts below regarding the points mentioned in the review checklist:

1. Installation: Does installation proceed as outlined in the documentation? 

- I was not able to install the software by following the installation instructions. I had to manually install the `mongiris` package following the installation instructions [here](https://gitlab.liris.cnrs.fr/fduchate/mongiris). It would be much convenient to use docker VM for setting up Predihood as already mentioned in one of the comments above. 
So, the installation needs to be checked and reviewed comprehensively. 
- After installing it, I was not able to run the Predihood software as mentioned [here](https://gitlab.liris.cnrs.fr/fduchate/mongiris) by executing `python3 main.py`.  Please follow the error log below and please correct me if I am following any step incorrectly. The error is possibly due to the missing `configs` directory as mentioned [here](https://gitlab.com/fduchate/predihood/-/blob/master/predihood/main.py#L24): 
```
(predihood) rajat@rajat-infinity:/mnt/d/JOSS Reviews/predihood/predihoodClone/predihood$ python3 main.py
WARNING:__main__:No parameter provided, loading default dataset configuration file (configs/config_hil.json)
Traceback (most recent call last):
  File "main.py", line 343, in <module>
    dataset_config = load_dataset_config(dataset_config_file)
  File "/mnt/d/JOSS Reviews/predihood/predihoodClone/predihood/utility_functions.py", line 537, in load_dataset_config
    with open(json_file_path) as data_file:
FileNotFoundError: [Errno 2] No such file or directory: 'configs/config_hil.json'
```
Though `python3 main.py datasets/hil/config.json` as described [here](https://gitlab.com/fduchate/predihood/-/tree/master/#using-another-dataset) worked perfectly.

2. Documentation: 
- Example usage: Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).

The examples worked like charm. I am adding some screenshots below in the Other section, verifying the functionality. Also, I really liked the documentation generated using `pdoc` but it could be more convenient if the entire document is hosted online (maybe using GitHub pages) as currently, it is required to open the `HTML` files manually from the local directory.

- Automated tests: Are there automated tests or manual steps described so that the functionality of the software can be verified?
    
    - The Predihood Documentation does not mention anything about the automated tests. Though, information about tests is present in the project repository README but not in the documentation. Please verify if I missed anything. 
    - While executing tests by `python3 tests.py`, I received the following error. It is possibly due to missing config and dataset files inside the `dataset` directory. I tried it again by copying the `csv` and `config` file from the `hil` [directory](https://gitlab.com/fduchate/predihood/-/tree/master/predihood/datasets/hil) to one level up [here](https://gitlab.com/fduchate/predihood/-/tree/master/predihood/datasets) in my working environment but the same error sustained. The authors are recommended to verify it. 
 ```
(predihood) rajat@rajat-infinity:/mnt/d/JOSS Reviews/predihood/predihoodClone/predihood$ python3 tests.py
test_add_assessment_to_file (__main__.TestCase) ... ok
test_address_to_city (__main__.TestCase) ... ok
test_address_to_code (__main__.TestCase) ... ok
test_get_classifier (__main__.TestCase) ... ok
test_get_most_frequent (__main__.TestCase) ... ok
test_indicator_full_to_short_label (__main__.TestCase) ... ERROR
test_indicator_short_to_full_label (__main__.TestCase) ... ERROR
test_intersection (__main__.TestCase) ... ok
test_set_classifier (__main__.TestCase) ... ok
test_signature (__main__.TestCase) ... ok
test_similarity (__main__.TestCase) ... ok
test_train_test_percentages (__main__.TestCase) ... ok
test_union (__main__.TestCase) ... ok
test_values_dataset (__main__.TestCase) ... ERROR

======================================================================
ERROR: test_indicator_full_to_short_label (__main__.TestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tests.py", line 43, in test_indicator_full_to_short_label
    short_label = indicator_full_to_short_label(full_label)
  File "/mnt/d/JOSS Reviews/predihood/predihoodClone/predihood/utility_functions.py", line 113, in indicator_full_to_short_label
    indicators = model.get_indicators_dict()
  File "/mnt/d/JOSS Reviews/predihood/predihoodClone/predihood/model.py", line 153, in get_indicators_dict
    list_indicators = db.find_all(db.collection_indicators)
AttributeError: 'NoneType' object has no attribute 'find_all'

======================================================================
ERROR: test_indicator_short_to_full_label (__main__.TestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tests.py", line 49, in test_indicator_short_to_full_label
    full_label = indicator_short_to_full_label(short_label)
  File "/mnt/d/JOSS Reviews/predihood/predihoodClone/predihood/utility_functions.py", line 132, in indicator_short_to_full_label
    indicators = model.get_indicators_dict()
  File "/mnt/d/JOSS Reviews/predihood/predihoodClone/predihood/model.py", line 153, in get_indicators_dict
    list_indicators = db.find_all(db.collection_indicators)
AttributeError: 'NoneType' object has no attribute 'find_all'

======================================================================
ERROR: test_values_dataset (__main__.TestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tests.py", line 145, in test_values_dataset
    dataset = pd.read_csv(filename)
  File "/mnt/d/JOSS Reviews/predihood/predihood/lib/python3.8/site-packages/pandas/io/parsers.py", line 688, in read_csv
    return _read(filepath_or_buffer, kwds)
  File "/mnt/d/JOSS Reviews/predihood/predihood/lib/python3.8/site-packages/pandas/io/parsers.py", line 454, in _read
    parser = TextFileReader(fp_or_buf, **kwds)
  File "/mnt/d/JOSS Reviews/predihood/predihood/lib/python3.8/site-packages/pandas/io/parsers.py", line 948, in __init__
    self._make_engine(self.engine)
  File "/mnt/d/JOSS Reviews/predihood/predihood/lib/python3.8/site-packages/pandas/io/parsers.py", line 1180, in _make_engine
    self._engine = CParserWrapper(self.f, **self.options)
  File "/mnt/d/JOSS Reviews/predihood/predihood/lib/python3.8/site-packages/pandas/io/parsers.py", line 2010, in __init__
    self._reader = parsers.TextReader(src, **kwds)
  File "pandas/_libs/parsers.pyx", line 382, in pandas._libs.parsers.TextReader.__cinit__
  File "pandas/_libs/parsers.pyx", line 674, in pandas._libs.parsers.TextReader._setup_parser_source
FileNotFoundError: [Errno 2] No such file or directory: '../predihood/datasets/data_density.csv'

----------------------------------------------------------------------
Ran 14 tests in 7.446s

FAILED (errors=3)
```

3. Software paper

- The software paper contains some spelling errors and typos in the repository. The authors are recommended to carefully review the software paper for the following issues:
    - Incorrect Bibtex for reference 2
       Replace `livehoods` -> `Livehoods`
    - Typo
       Section: Methodology (citing from the paper)
`Predihood provides the following functionnalities: - adding new neighbourhoods and indicators to describe them; - predict the environment of a neighbourhood by configuring and usingpredefined algorithms; - adding new predictive algorithms.`
        - Replace `functioannalities` -> `functionalities`
        - Bulleted points are not rendered properly 

       Section: Adding new data
`It includes about 50,000 neighbourhoods with 640 indicators, and 270 neighbouhoods were`
        - `and 270 neighbouhoods` -> `and 270 neighbourhoods`

       Section: Predicting environment
        - `optionnaly` -> `optionally`

4. Other
Below are the screenshots:

    - Main screen
        ![1](https://user-images.githubusercontent.com/21292545/110232195-5a086200-7f42-11eb-86cf-a8cc31345ea8.png)
    -  Cartographic Interface of Predihood
        ![2](https://user-images.githubusercontent.com/21292545/110232196-5bd22580-7f42-11eb-8ebd-4eff9f035c70.png)
    -   Results obtained after executing Random Forest Classifier
        ![3](https://user-images.githubusercontent.com/21292545/110232198-5d035280-7f42-11eb-9375-d36822796da1.png)
    -   Visualization of Confusion Matrix
        ![4](https://user-images.githubusercontent.com/21292545/110232199-5ffe4300-7f42-11eb-9775-c3cc4671b298.png)

     
My remarks:
I would like to congratulate the authors for their efforts and for developing a neat software package. I envisage that Predihood would be very much useful and finds extended applications in various domains along with the core objective of predicting insights about neighborhoods. I would recommend the authors review the above feedback, especially the existing issues with the installation. 
Kind regards,
Rajat

**Comment by galessiorob** [comment]
Thank you so much for your thoughtful and thorough review @omshinde! 

@fduchatea I'd encourage you to consider using a Docker container to avoid these installation issues, please take note of the other issues all the reviewers have raised and let me know if you have any questions or if I can support you in any way.

**Comment by nuest** [comment]
I ran into the same installation problems and reported here: https://gitlab.com/fduchate/predihood/-/issues/1#note_526210425

Most other comments I had seem to be addressed, though it's not that easy to get into the project again. As I'd probably have to start my review from the beginning, I'd prefer to do this only once and would prefer to wait until

1. _all_ installation problems are resolved, or mitigated with a docker-compose configuration.
2. the paper itself has been updated (at least the project URL is not on GitLab - or do the authors plan to maintain both clones?)
3. tests have been fixed (see below).


<details>
<summary>Details</summary>

```
(joss-predihood) daniel@nuest ~/git/predihood/predihood [master]$ python tests.py 
test_add_assessment_to_file (__main__.TestCase) ... ok
test_address_to_city (__main__.TestCase) ... ok
test_address_to_code (__main__.TestCase) ... ok
test_get_classifier (__main__.TestCase) ... ok
test_get_most_frequent (__main__.TestCase) ... ok
test_indicator_full_to_short_label (__main__.TestCase) ... ERROR
test_indicator_short_to_full_label (__main__.TestCase) ... ERROR
test_intersection (__main__.TestCase) ... ok
test_set_classifier (__main__.TestCase) ... ok
test_signature (__main__.TestCase) ... ok
test_similarity (__main__.TestCase) ... ok
test_train_test_percentages (__main__.TestCase) ... ok
test_union (__main__.TestCase) ... ok
test_values_dataset (__main__.TestCase) ... ERROR

======================================================================
ERROR: test_indicator_full_to_short_label (__main__.TestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tests.py", line 43, in test_indicator_full_to_short_label
    short_label = indicator_full_to_short_label(full_label)
  File "/home/daniel/git/predihood/predihood/utility_functions.py", line 113, in indicator_full_to_short_label
    indicators = model.get_indicators_dict()
  File "/home/daniel/git/predihood/predihood/model.py", line 153, in get_indicators_dict
    list_indicators = db.find_all(db.collection_indicators)
AttributeError: 'NoneType' object has no attribute 'find_all'

======================================================================
ERROR: test_indicator_short_to_full_label (__main__.TestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tests.py", line 49, in test_indicator_short_to_full_label
    full_label = indicator_short_to_full_label(short_label)
  File "/home/daniel/git/predihood/predihood/utility_functions.py", line 132, in indicator_short_to_full_label
    indicators = model.get_indicators_dict()
  File "/home/daniel/git/predihood/predihood/model.py", line 153, in get_indicators_dict
    list_indicators = db.find_all(db.collection_indicators)
AttributeError: 'NoneType' object has no attribute 'find_all'

======================================================================
ERROR: test_values_dataset (__main__.TestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tests.py", line 145, in test_values_dataset
    dataset = pd.read_csv(filename)
  File "/home/daniel/git/predihood/.venv/joss-predihood/lib/python3.8/site-packages/pandas/io/parsers.py", line 688, in read_csv
    return _read(filepath_or_buffer, kwds)
  File "/home/daniel/git/predihood/.venv/joss-predihood/lib/python3.8/site-packages/pandas/io/parsers.py", line 454, in _read
    parser = TextFileReader(fp_or_buf, **kwds)
  File "/home/daniel/git/predihood/.venv/joss-predihood/lib/python3.8/site-packages/pandas/io/parsers.py", line 948, in __init__
    self._make_engine(self.engine)
  File "/home/daniel/git/predihood/.venv/joss-predihood/lib/python3.8/site-packages/pandas/io/parsers.py", line 1180, in _make_engine
    self._engine = CParserWrapper(self.f, **self.options)
  File "/home/daniel/git/predihood/.venv/joss-predihood/lib/python3.8/site-packages/pandas/io/parsers.py", line 2010, in __init__
    self._reader = parsers.TextReader(src, **kwds)
  File "pandas/_libs/parsers.pyx", line 382, in pandas._libs.parsers.TextReader.__cinit__
  File "pandas/_libs/parsers.pyx", line 674, in pandas._libs.parsers.TextReader._setup_parser_source
FileNotFoundError: [Errno 2] No such file or directory: '../predihood/datasets/data_density.csv'

----------------------------------------------------------------------
Ran 14 tests in 4.412s

FAILED (errors=3)
```

</details>

**Comment by fduchatea** [comment]
@omshinde  Thank you for the review. We are working on facilitating the installation with docker, then we will address your comments.

@nuest We keep you updated when the docker installation is working. 
We are not maintaining both git repositories (the liris repository is still up, but only with a short readme file referring to the gitlab repository) and we will update the paper accordingly.

**Comment by fduchatea** [comment]
@galessiorob @omshinde @nuest @martinfleis: we are happy to tell you that Predihood now runs using Docker.
Following are the commands for the first run:
```
git clone https://gitlab.liris.cnrs.fr/fduchate/mongiris
git clone https://gitlab.com/fduchate/predihood.git
cd predihood/  # or equivalent command on non-bash systems
docker-compose up
```
Note that the first run takes a few minutes (due to the data loading part), but next ones (using only `docker-compose up`) are fast.

The README file has been updated accordingly (we also kept the manual installation).

As Rajat suggested, documentation includes tests and is hosted online: [https://nellybarret.gitlab.io/documentation-for-predihood](https://nellybarret.gitlab.io/documentation-for-predihood)

We have rewritten some parts of the paper to better focus on the tool's functionalities (and corrected typos mentioned by Rajat).

Thank you @omshinde for your review, and all for your comments so far!

**Comment by omshinde** [comment]
@fduchatea: That's amazing to know, thanks for the update! I will try to review the changes by Tuesday.

Kind regards,
Rajat

**Comment by martinfleis** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02805/joss.02805/10.21105.joss.02805.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02805/joss.02805/10.21105.joss.02805.pdf) :page_facing_up: :point_left:

**Comment by martinfleis** [comment]
Thansk for all the changes @fduchatea. I was finally able to run `predihood` using `docker-compose` and it worked like a charm. Therefore, I am able to finalise my review. 

I believe that there is a scope in predihood's approach to neighbourhood assessment. In my understanding, the app is aimed at users without a technical knowledge allowing them to deploy ML models manually, hence the GUI. I think that with docker, this may work. That said, I am surely not the intended user.

The software is well documented with nice examples, its code is clean and the software paper sufficiently covers all required aspects. I have just two minor notes which should be ideally resolved.

When I download a result of a prediction as xls, I get the following warning from Excel. Not sure what is causing it. It is emitted by Excel for Mac 16.47.1. Also, the paper mentions that the result is downloaded as CSV but the file is XLS.

<img width="292" alt="Screenshot 2021-03-28 at 11 50 12" src="https://user-images.githubusercontent.com/36797143/112751217-a6384680-8fc4-11eb-9356-111642d09b0f.png">

The table in figure 1 in the paper should not contain red lines underlining spelling issues.

None of these issues block the publication. I can recommend this paper for publication in JOSS.

**Comment by galessiorob** [comment]
Great work @fduchatea!🎉🎉🎉 

Thank you @martinfleis for the latest assessment.

@fduchatea want to fix that last but so that the file types are consistent?

@omshinde @nuest @jdalzatec could you all do it what will likely be the last review? 

Thanks for the patience and dedication.

**Comment by omshinde** [comment]
Thanks, @fduchatea, for the changes. The paper is now updated, and the typos are fixed. The `docker` based installation is now very smooth, and I am sure it will definitely enhance the usage of Predihood. Also, the hosted documentation looks great. Thanks for addressing all the comments.

My final assessment: 

Congratulations to the authors for their efforts. I foresee that Predihood finds applications in various domains along with the core objective of predicting insights about neighborhoods. The installation is smooth, coding is neat, the software is well-documented, and the software paper succinctly describes the need and feasibility for the development of Predihood. I have checked all the boxes in my review list, and I have a full recommendation for Predihood as a publication in JOSS.

Thanks to @galessiorob for all the help during the review cycle.

Kind regards,
Rajat

**Comment by fduchatea** [comment]
@omshinde @martinfleis @galessiorob : Thank you for your feedback and positive reviews!

We have corrected the XLS output format in the paper. 
About the warning: it is due to the converter API (from HTML to XLS). We are 2 MacOS users in the team: one also uses Excel and has the same warning as Martin while I use Libre Office to read the XLS file and I do not have any warning. It seems we cannot do much about it.

**Comment by martinfleis** [comment]
> It seems we cannot do much about it.

Maybe a short note in the docs that this may happen? So the user knows it is expected behaviour and there's nothing to be worried about.

**Comment by fduchatea** [comment]
> Maybe a short note in the docs that this may happen? So the user knows it is expected behaviour and there's nothing to be worried about.

Done (in the `predict.py` file, and on the package documentation). Will be soon on the online documentation too (needs a manual intervention).

**Comment by galessiorob** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02805/joss.02805/10.21105.joss.02805.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02805/joss.02805/10.21105.joss.02805.pdf) :page_facing_up: :point_left:

**Comment by galessiorob** [comment]
@fduchatea Amazing work! Thanks for adding @martinfleis and @omshinde's recommendations.

I have opened a PR with edit suggestions [here](https://gitlab.com/fduchate/predihood/-/merge_requests/4), aside from those I think it would be good to label your figures too.

Let me know when you have taken a look, after that, I think we'll be ready to publish.

**Comment by fduchatea** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02805/joss.02805/10.21105.joss.02805.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02805/joss.02805/10.21105.joss.02805.pdf) :page_facing_up: :point_left:

**Comment by fduchatea** [comment]
@galessiorob Thank you for improving the paper!
We have checked the paper and added labels to the figures, and whedon has produced the latest version (post above).

**Comment by nuest** [comment]
@galessiorob I've taken another look and updated my checklist above and a few minor comments. As the other reviewers have already given their :+1: I leave it to you to judge if any of these should hold up publication or are left for the authors to clarify in the README later.

- Because I cloned the repo to my "regular" folder with _many_ git repositories, all those files became part of the Docker context. I suggest to adjust the instructions accordingly and say "clone the repositories into a _new directory_".
- I could run the tests manually as described in the README. I _strongly_ encourage the authors to integrate these tests in a CI workflow.
- For the scholarly efforts, I would feel more comfortable having checked the box if the authors stress even more what [@martinfleis pointed out](https://github.com/openjournals/joss-reviews/issues/2805#issuecomment-808886509): that predihood makes it very easy, even for non programmers, to use predictive algorithms;

---------

Congratulations to the authors and other reviewers for the improvements made during this review - I'm happy to see the paper and software ready to be published in JOSS.

**Comment by fduchatea** [comment]
@nuest Thank you for the feedback !
We have added a sentence in the README to clarify installation ("both cloned repositories should be placed into a new (empty) directory").
We agree that Predihood is more targeted towards non-IT researchers (e.g., in social sciences, history, biology). We have added a sentence in README introduction ("It makes it very easy, even for non programmers, to use predictive algorithms") and another one in the paper ("Predihood mainly targets non-programmers users (e.g., researchers in social sciences or history) due to its simplicity for running and configuring predictive algorithms").
For the CI workflow, maybe we could let this for later (especially if several contributors join the project) as it would require a few days/weeks to add this feature.

**Comment by galessiorob** [comment]
Thanks for adding those final suggestions @nuest and @fduchatea - I agree that you can note the CI workflow as future improvements.

@jdalzatec could you please update your checklist? I believe this is the last bit we need to complete to get this paper published.

**Comment by jdalzatec** [comment]
Hi @fduchatea and authors!
cc: @galessiorob

System settings while reviewing:
OS - Manjaro Linux
Web browser: Firefox

Thank you for having me review this great application. I found this tool very interesting and exciting to use. It gives you very good information and it is really easy to use. I am sure people are going to use it considering that is both useful and fast. Results are obtained quickly and you can tune a lot of parameters according to your needs. The code was written using some goods design patterns and it is relatively easy to read. I really enjoyed reviewing this. I just have some additional comments, which are listed below. I would like to congratulate the authors for their effort in doing this great tool.

Best regards,
Juan

### Unexpected behaviors:

1. When I look for a neighborhood (e.g. Lyon) and I click over, I got an error when trying to select the algorithm. I attach the evidence.
![image](https://user-images.githubusercontent.com/2148473/115125967-55978600-9f91-11eb-90b7-4860dbdaa857.png)
![image](https://user-images.githubusercontent.com/2148473/115125979-66e09280-9f91-11eb-842b-7e11fd53d5c6.png)

2. Main button to start prediction should be set as a primary button:
![image](https://user-images.githubusercontent.com/2148473/115126005-9abbb800-9f91-11eb-8599-935f7f6ad8bb.png)

3. When you click on "Predict selected neighborhoods", the cursor change to a loading state. However, controls are not disabled and you can change parameters or even click again on the "Predict selected neighborhoods" button, which starts another prediction. In that way, you get confused if your prediction started correctly or not.

4. It is not very clear what this control is for:
![image](https://user-images.githubusercontent.com/2148473/115126091-4f55d980-9f92-11eb-87cf-900ba61610e7.png)

5. Cursor for the blue download button was set as a move one (a hand). It should be as an actionable cursor (a link for example).

6. On the page for "tune algorithms and to test them", when you "Train, test and evaluate", the cursor changes to a loading state. However, controls are not disabled and, therefore, you can change parameters or click on the button again to start another training. In that way, you could feel as the action was not executed.

### Software

* It would be great if the authors enforce a guide style. Flake8 is a very useful tool.
* Integrate a command (maybe a script) to run tests outside the container.
* It would be great if the authors increase the coverage for the software (currently at 21%):
![image](https://user-images.githubusercontent.com/2148473/115126448-b3799d00-9f94-11eb-8919-71745b27c450.png)

**Comment by galessiorob** [comment]
Thanks so much for your review @jdalzatec!

@fduchatea could you take a quick look at the error and assess how large of a fix this would be? It would also be useful to open issues for the button suggestions but I don't think this should be a blocker for us to publish. Thanks again for your tenacity on this paper.

**Comment by fduchatea** [comment]
@jdalzatec : Thank you for the review!

Following are some answers to the mentioned points:
1. (error) in the tooltip, you should select an algorithm for predicting (the predict process is automatically run after a change in the list). It seems that you have selected "---" and this has caused an error. 
When selecting an algorithm in the list for this neighbourhood, prediction works (as shown on the screenshot below). On our side, we added a test to check that the value is valid to avoid this error.
<img width="922" alt="Capture d’écran 2021-04-27 à 10 02 29" src="https://user-images.githubusercontent.com/54939958/116207389-2080f580-a740-11eb-93a6-4fb0ec30d328.png">

2. fixed button color
3. todo
4. We agree that it is not so clear, but we did not find a better label. The idea is to be able to display automatically neighbourhoods that are on the current map view. Howevern this is not possible with low zoom levels (too many neighbourhoods, so it takes much time, and it is not very useful). Thus the automatic display of neighbourhoods in the map is activated when the zoom level is above 12 (default value in the interface). This value can be changed directly in the interface if needed. During our tests, users find this feature very useful, but the drawback is that zomming in/out or moving on the map reloads the neighbourhoods on the visible part of the map (which does not help for working on an area, trying to predict adjacent neighbourhoods). Thus, we implemented a "lock" (the lock button close to the zoom) so that displayed neighbourhoods do not change when the lock is on. The screenshot below illustrates this feature, we have unlocked the display, zoomed up to level 12 (and all neighbourhoods on the visible part of the map were displayed), and then we locked the displaying (and we zoomed out for a better picture, so that we see the "map rectangle" from which all neighbourhoods were extracted). On such map, researchers can work on all displayed neighbourhoods (checking their information, predicting, etc) without having to perform multiple searches.
<img width="1380" alt="Capture d’écran 2021-04-27 à 10 56 06" src="https://user-images.githubusercontent.com/54939958/116215089-a6ed0580-a747-11eb-9c47-bc74998a992f.png">

5. fixed the cursor to pointer
6. todo

Software:
1. todo
2. running the tests outside the container only requires a single python command (`python3 tests.py`), but also requires launching the local database. It is not possible to automate the start of the local database (very different from one user to another) so a script to run the test would only contain the single python command. We think it is not worth to define such script (besides, several scripts would be needed, at least a shell script and a DOS script).
3. todo

We will take into account the remaining points as soon as possible, and keep you updated.

**Comment by galessiorob** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02805/joss.02805/10.21105.joss.02805.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02805/joss.02805/10.21105.joss.02805.pdf) :page_facing_up: :point_left:

**Comment by galessiorob** [comment]
Thanks for your responses @fduchatea! I think we're good to publish as is and you can work on the remaining fixes since the core functionality is there and the readme is pretty thorough.

Could you proofread this [final version](https://github.com/openjournals/joss-papers/blob/joss.02805/joss.02805/10.21105.joss.02805.pdf), please? 🙏

Once you greenlight it we can accept it and publish it 🎉

**Comment by fduchatea** [comment]
@galessiorob : the final version of the paper is fine.
@jdalzatec : we have added fieldsets to disable fields when predicting (on both interfaces, point 3 and 6). We have also used SonarLint to improve code quality.

Thank you all for your fruitful feedback!

**Comment by galessiorob** [comment]
@fduchatea last thing we need from you is to deposit your software in Zenodo and post the DOI here. After that, we can set the final version, generate a final proof, and have an Editor in Chief do the publishing. 🚀

**Comment by fduchatea** [comment]
@galessiorob Here is the Zenodo DOI: [http://doi.org/10.5281/zenodo.4737729](http://doi.org/10.5281/zenodo.4737729) 🚀

**Comment by galessiorob** [comment]
@whedon set v1.1 as version

**Comment by whedon** [comment]
OK. v1.1 is the version.

**Comment by galessiorob** [comment]
@whedon set 10.5281/zenodo.4737729 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.4737729" target="_blank">10.5281/zenodo.4737729</a> is the archive.

**Comment by galessiorob** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02805/joss.02805/10.21105.joss.02805.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02805/joss.02805/10.21105.joss.02805.pdf) :page_facing_up: :point_left:

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1109/socialcom.2013.17 is OK
- 10.5220/0009885702940301 is OK
- 10.2188/jea.JE20130084 is OK
- 10.1136/bjsm.2009.058701 is OK
- 10.3390/su10030575 is OK
- 10.1098/rsbl.2018.0082 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by galessiorob** [comment]
@fduchatea could you please give this last version a final proofread? An editor in chief will take it from here and get it published now ✨

@jdalzatec, @omshinde, @nuest, @martinfleis MASSIVE THANKS TO YOU ALL! This paper and software went from being good to great, appreciate your time and effort ❤️

**Comment by arfon** [comment]
This is looking good folks. Could you please make sure you cite Scikit learn and Numpy here please (and any other software packages you mention)? You can learn how here:

- https://scikit-learn.org/stable/about.html#citing-scikit-learn
- https://numpy.org/citing-numpy/

**Comment by fduchatea** [comment]
@whedon  generate pdf

**Comment by whedon** [comment]
I'm sorry human, I don't understand that. You can see what commands I support by typing:

```
@whedon commands
```

**Comment by fduchatea** [comment]
@whedon generate pdf

**Comment by arfon** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1038/s41586-020-2649-2 is OK
- 10.5281/zenodo.4161697 is OK
- 10.1109/MCSE.2007.55 is OK
- 10.21105/joss.03021 is OK
- 10.1109/socialcom.2013.17 is OK
- 10.5220/0009885702940301 is OK
- 10.2188/jea.JE20130084 is OK
- 10.1136/bjsm.2009.058701 is OK
- 10.3390/su10030575 is OK
- 10.1098/rsbl.2018.0082 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/2292

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/2292, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by arfon** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1391384196237692931) 👈 🐦🐦🐦

**Comment by arfon** [comment]
@jdalzatec, @omshinde, @nuest, @martinfleis - many thanks for your reviews here and to @galessiorob for editing this submission. JOSS relies up the volunteer efforts of people like you, and we simply wouldn't be able to do this without you! ✨ 

@fduchatea - your paper is now accepted and published in JOSS :zap::rocket::boom:

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.02805/status.svg)](https://doi.org/10.21105/joss.02805)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.02805">
  <img src="https://joss.theoj.org/papers/10.21105/joss.02805/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.02805/status.svg
   :target: https://doi.org/10.21105/joss.02805
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.02805/status.svg)](https://doi.org/10.21105/joss.02805)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss

**Comment by omshinde** [comment]
@arfon @galessiorob Thank you for the opportunity.

@fduchatea Congratulations on the amazing work! :)

**Comment by fduchatea** [comment]
@arfon Thank you for the publication! 🎉 
On the publication webpage (https://joss.theoj.org/papers/10.21105/joss.02805), the link to software repository is not correct (it is the initial link on liris.cnrs.fr, with issues regarding possibility of collaboration by anyone). The correct link is [https://gitlab.com/fduchate/predihood](https://gitlab.com/fduchate/predihood)

Thank you very much @galessiorob and all reviewers for helping us improving our submission!
